File: src/harness_mcp/connectors.py
import requests
import argparse
import json
import os
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

def list_connectors(connector_names=None, connector_identifiers=None, types=None):
    """
    List connectors from Harness API
    
    Args:
        connector_names: Optional list of connector names
        connector_identifiers: Optional list of connector identifiers
        types: Optional list of connector types
    
    Returns:
        API response as JSON
    """
    url = "https://app.harness.io/ng/api/connectors/listV2"
    
    # Query parameters
    params = {
        "accountIdentifier": "wFHXHD0RRQWoO8tIZT5YVw",
        "orgIdentifier": "default",
        "projectIdentifier": "nitisha",
        "getDefaultFromOtherRepo": "true",
        "getDistinctFromBranches": "true",
        "onlyFavorites": "false",
        "pageIndex": "0",
        "pageSize": "50",
        "sortOrders": "orderType=ASC"
    }
    
    # Headers
    headers = {
        "Content-Type": "application/json",
        "x-api-key": os.getenv("HARNESS_API_KEY")
    }
    
    # Request body
    payload = {
        "categories": ["CLOUD_PROVIDER"],
        "filterType": "Connector"
    }
    
    # Add optional parameters if provided
    if connector_names:
        payload["connectorNames"] = connector_names
    
    if connector_identifiers:
        payload["connectorIdentifiers"] = connector_identifiers
    
    if types:
        payload["types"] = types
    
    # Make the request
    response = requests.post(url, params=params, headers=headers, json=payload)
    
    return response.json()

def main():
    parser = argparse.ArgumentParser(description="List connectors from Harness API")
    parser.add_argument("--names", nargs="+", help="List of connector names")
    parser.add_argument("--identifiers", nargs="+", help="List of connector identifiers")
    parser.add_argument("--types", nargs="+", help="List of connector types")
    args = parser.parse_args()
    
    response = list_connectors(
        connector_names=args.names,
        connector_identifiers=args.identifiers,
        types=args.types
    )
    
    print(json.dumps(response, indent=2))

if __name__ == "__main__":
    main()

File: pkg/harness/connectors.go
package harness

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

// Base URL for Harness API
const baseURL = "https://app.harness.io/ng/api"

// Default account, org, and project identifiers
// In a real implementation, these would be configurable
// TODO: remove these
const (
	defaultAccountID = "wFHXHD0RRQWoO8tIZT5YVw"
	defaultOrgID     = "default"
	defaultProjectID = "nitisha"
)

// ConnectorClient provides methods to interact with Harness API for connectors
type ConnectorClient struct {
	httpClient *http.Client
	baseURL    string
}

// NewConnectorClient creates a new connector client
func NewConnectorClient() *ConnectorClient {
	return &ConnectorClient{
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
		baseURL: baseURL,
	}
}

// ListConnectorRequest represents the request to list connectors
type ListConnectorRequest struct {
	Categories           []string `json:"categories,omitempty"`
	ConnectorNames       []string `json:"connectorNames,omitempty"`
	ConnectorIdentifiers []string `json:"connectorIdentifiers,omitempty"`
	Types                []string `json:"types,omitempty"`
	FilterType           string   `json:"filterType"`
}

// ListConnectors retrieves connectors from the Harness API
func (c *ConnectorClient) ListConnectors(ctx context.Context, connectorNames []string, connectorIDs []string, types []string) ([]byte, error) {
	// Get API key from context
	apiKey, err := GetApiKeyFromContext(ctx)
	if err != nil {
		return nil, err
	}

	// Construct the query parameters
	params := url.Values{}
	params.Add("accountIdentifier", defaultAccountID)
	params.Add("orgIdentifier", defaultOrgID)
	params.Add("projectIdentifier", defaultProjectID)
	params.Add("getDefaultFromOtherRepo", "true")
	params.Add("getDistinctFromBranches", "true")
	params.Add("onlyFavorites", "false")
	params.Add("pageIndex", "0")
	params.Add("pageSize", "50")
	params.Add("sortOrders", "orderType=ASC")

	// Create the request body
	reqBody := ListConnectorRequest{
		Categories:           []string{"CLOUD_PROVIDER"},
		ConnectorNames:       connectorNames,
		ConnectorIdentifiers: connectorIDs,
		Types:                types,
		FilterType:           "Connector",
	}

	// Marshal the request body to JSON
	reqBodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request body: %w", err)
	}

	// Create a new HTTP request
	url := fmt.Sprintf("%s/connectors/listV2?%s", c.baseURL, params.Encode())
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewBuffer(reqBodyBytes))
	if err != nil {
		return nil, fmt.Errorf("failed to create HTTP request: %w", err)
	}

	// Set headers
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", apiKey)

	// Make the HTTP request
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make HTTP request: %w", err)
	}
	defer resp.Body.Close()

	// Read the response body
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	// Check if the response status code is not 2xx
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP request failed with status code %d: %s", resp.StatusCode, string(body))
	}

	// Return the raw JSON response
	return body, nil
}

// ListConnectorsTool creates the list-connector tool
func ListConnectorsTool(client *ConnectorClient) (mcp.Tool, server.ToolHandlerFunc) {
	tool := mcp.NewTool("list-connector",
		mcp.WithDescription("List available connectors"),
		mcp.WithString("connector_type",
			mcp.Description("Type of connector to filter by")),
		mcp.WithArray("connector_names",
			mcp.Description("List of connector names to filter by"),
			mcp.Items(map[string]interface{}{"type": "string"})),
		mcp.WithArray("connector_ids",
			mcp.Description("List of connector IDs to filter by"),
			mcp.Items(map[string]interface{}{"type": "string"})),
	)

	handler := func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		// Extract parameters
		connectorType, err := OptionalParam[string](request, "connector_type")
		if err != nil {
			return nil, err
		}

		connectorNames, err := OptionalStringArrayParam(request, "connector_names")
		if err != nil {
			return nil, err
		}

		connectorIDs, err := OptionalStringArrayParam(request, "connector_ids")
		if err != nil {
			return nil, err
		}

		// Prepare types array if connector_type is specified
		var types []string
		if connectorType != "" {
			types = []string{connectorType}
		}

		// Call Harness API
		jsonResponse, err := client.ListConnectors(ctx, connectorNames, connectorIDs, types)
		if err != nil {
			return nil, err
		}

		// Return the raw JSON response as a text result
		return mcp.NewToolResultText(string(jsonResponse)), nil
	}

	return tool, handler
}
